### PR TITLE
Bump spring-boot 2.7.18 to 3.2.1

### PR DESCRIPTION
**Pull request by tesshucom**
Prerequisites: #839

## Overview

Spring Boot will be majorly upgraded to 3.2.x. Along with this, the Servlet version will be updated to 6.0, and the App Server will also undergo a major version upgrade. Related libraries will also be updated.

## Background (About Spring Boot 3.x)

<details>
<summary>Major version upgrade of Spring Boot is inevitable</summary>

This is a diagram showing the platform-related life cycle drawn in 2023. 

Jpsonic has already raised the minimum Java version from 11 to 17. The next thing we need to do is update Spring Boot. If this is not done, there will be disadvantages such as related libraries not being able to be updated. (e.g. As a known issue, logging library support is already EOL.)

image

</details>

Originally, the deadline for this was Nov 2023. The planned delay was intentional, and the reason was that we wanted to have the same servlet versions for Tomcat and Jetty.

 - [Apache Tomcat Versions](https://tomcat.apache.org/whichversion.html)
 - [Eclipse Jetty Versions](https://eclipse.dev/jetty/)

Currently, the latest practical Sevlet version is 6.0 (Practical = Latest version available for Java17). To meet this requirement, we had to wait for Jetty 12.0.x support from Spring. In order to migrate with the minimum number of steps, it would be efficient to postpone the bump from Spring Boot 2.7.x to 3.0 and use 3.2 as the landing point. The spec for the next version of Jpsonic will be as follows:

Jpsonic | Jetty | Tomcat | Servlet Spec | Spring Boot | Namespace | Java
-- | -- | -- | -- | -- | -- | --
v114 | 12.0.x | 10.1.x | 6 | 3.2 | jakarta.servlet | 17, 21

## Goal

The review items for this Pull Request are generally as follows.

 - Spring Boot, servlet related Libs will be bumped
 - Verifying that the test library works properly. Also confirm that existing tests are working properly. 
 - Confirm that the Tomcat and Jetty embed configurations run correctly.
 - Confirme that war deployment is possible in Tomcat and Jetty.
 - Log in with Docker configuration and confirm that the affected functions are working properly.
 - Verify Subsonic Apps work as before

## Non-Goal

There will likely be quite a few implementation fixes involved. Therefore, some items other than those indicated by 'Goal' will be resolved independently in subsequent pull requests. Or, in order to reach users more quickly, it may be split into those that will be fixed in the release version and those that will be followed up with hotfixes.

The following items may be added during review 🙂🙃

#### Not works, so should be rewritten

 - Supports Servlet 6.0 for UPnP service 🥰 < IMPORTANT!
   - **This pull request will cause UPnP to temporarily stop working.**
   - Since there is no library compatible with Servlet 6.0, you will need to apply a patch to the 3rd-party library. 
   - jupnp may be a candidate.
 - Replacement of deprecated code between Spring Boot 3.0 and 3.2 
   - **This pull request will not confirm that it works.**
   - Migrate to fileupload2
     - This pull request tentatively rewrites it, but it will all need to be reimplemented and tested. This is because they are supported by the servlet, but unlike the previous process by commns, there are some interfaces that have disappeared. (progress monitoring)

#### Works, but should be rewritten

Required for release ...

 - Error Page
 - Replacement of AuthenticationFilter
   - The priority now is to understand the whole situation. Black magic is being used temporarily 😉

#### Works, but rewriting is preferable

This may be a feature release or hotfix, rather than a major release 😗

 - DWR related implementation
   - directwebremoting/dwr is still being used. However, temporarily, Patch Libs with rewritten naming space for Servlet6 is being used. Although it works, it would be better to replace it with Spring's standard configuration.
 - JWT related implementation
   - It might be a good idea to follow Spring's new design, although not right away. If anything, it may be better to do this at the same time as adding features such as extending deadlines. That doesn't mean it has to be done right away.
 - Replacement of deprecated code between Spring Boot 3.0 and 3.2 
   - Spring Theme related
     - That means doing it with CSS. This is not important, so we can leave it for later!

## Details

<details>
<summary>Breakdown of this pull request</summary>

 - Replacement of the last Derecated presented in Spring Boot 2.7.x.
   - Migrate WebSecurityConfigurerAdapter to SecurityFilterChain
     - **Contains black magic.**
   - Migrate authorizeRequests to authorizeHttpRequests
     - Corrections that could not be rewritten due to a bug and were left until the end. Not that important
 - Change! Change! Change! 
   - Bump spring-boot 2.7.18 to 3.2.1
     - There was a storm, but most of them were in line with the typical Spring-migration process.
     - There is one situation unique to this project. It's tag library migration.
       - Existing tag library has been made into a sub-project as 'airsonic-taglibs'.
         - To clarify dependencies of existing tag libraries.
         - The loading method of tag libraries differs depending on Apps Server and servlet version. A relatively obvious and reliable means is to isolate them into Jars instead of handling them within the main Project.
         - One didn't work so it was rewritten. Some depended on ancient commons, so they were partially copied. Leave everything else as is. It is not considered best to rewrite everything elegantly here.
    - Basically, there are no changes to the test content. Change the syntax to match the new API, etc. The code with relatively many differences is due to the Mockit specification change. In the previous ArgumentCapture, when the target was an Array, the way to represent the captured value was somewhat ambiguous. (Although it was usable for the time being.) The latest version can be written properly and accurately.
 - Replacement of deprecated code between Spring Boot 3.0 and 3.2 (Something of low difficulty. )
   - Migrate SimpleUrlHandlerMapping
   - Migrate ListenableFuture 
   - Migrate NestedServletException
   - Migrate apache.http.client to client5 

</details>

### CODE REVIEW

File: airsonic-taglibs/src/main/resources/META-INF/sub.tld [hunk 1]
```diff
@@ -1,13 +1,13 @@
 <?xml version="1.0" encoding="UTF-8" ?>
 
-<taglib xmlns="http://java.sun.com/xml/ns/j2ee"
-        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-        xsi:schemaLocation="http://java.sun.com/xml/ns/j2ee http://java.sun.com/xml/ns/j2ee/web-jsptaglibrary_2_0.xsd"
-        version="2.0">
+<taglib xmlns="http://java.sun.com/xml/ns/javaee"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-jsptaglibrary_2_1.xsd"
+    version="2.1">
```
Comment (tesshucom) [inline airsonic-taglibs/src/main/resources/META-INF/sub.tld:6]: Check taglib version.

File: jpsonic-main/src/main/java/com/tesshu/jpsonic/ajax/CoverArtService.java [hunk 2]
```diff
@@ -120,17 +121,18 @@ private void saveCoverArt(String pathString, String url) throws ExecutionExcepti
         backup(newCoverFile, Path.of(pathString, "cover." + suffix + ".backup"));
 
         try (CloseableHttpClient client = HttpClients.createDefault()) {
-            RequestConfig requestConfig = RequestConfig.custom().setConnectTimeout(2000).setSocketTimeout(2000).build();
+
+            RequestConfig requestConfig = RequestConfig.custom().setConnectionRequestTimeout(Timeout.ofSeconds(2))
+                    .setResponseTimeout(Timeout.ofSeconds(2)).build();
```
Comment (tesshucom) [inline jpsonic-main/src/main/java/com/tesshu/jpsonic/ajax/CoverArtService.java:126]: Use SocketTimeout.

File: jpsonic-main/src/main/java/com/tesshu/jpsonic/ajax/LyricsService.java [hunk 3]
```diff
@@ -128,12 +119,12 @@ private LyricsInfo parseSearchResult(String xml) throws ExecutionException {
     }
 
     private String executeGetRequest(String url) throws IOException {
-        RequestConfig requestConfig = RequestConfig.custom().setConnectTimeout(15_000).setSocketTimeout(15_000).build();
+        RequestConfig requestConfig = RequestConfig.custom().setConnectionRequestTimeout(Timeout.ofSeconds(15))
+                .setResponseTimeout(Timeout.ofSeconds(15)).build();
```
Comment (tesshucom) [inline jpsonic-main/src/main/java/com/tesshu/jpsonic/ajax/LyricsService.java:123]: Use SocketTimeout.

File: jpsonic-main/src/main/java/com/tesshu/jpsonic/controller/ProxyController.java [hunk 4]
```diff
@@ -58,13 +58,15 @@ public ModelAndView handleRequest(HttpServletRequest request, HttpServletRespons
             throws ServletRequestBindingException, ExecutionException {
         String url = ServletRequestUtils.getRequiredStringParameter(request, Attributes.Request.URL.value());
 
-        RequestConfig requestConfig = RequestConfig.custom().setConnectTimeout(15_000).setSocketTimeout(15_000).build();
+        RequestConfig requestConfig = RequestConfig.custom().setConnectionRequestTimeout(Timeout.ofSeconds(15))
+                .setResponseTimeout(Timeout.ofSeconds(15)).build();
```
Comment (tesshucom) [inline jpsonic-main/src/main/java/com/tesshu/jpsonic/controller/ProxyController.java:62]: Use SocketTimeout.

File: jpsonic-main/src/main/java/com/tesshu/jpsonic/service/VersionService.java [hunk 5]
```diff
@@ -267,14 +266,13 @@ private void readLatestVersion() throws IOException {
             if (LOG.isDebugEnabled()) {
                 LOG.debug("Starting to read latest version");
             }
-            RequestConfig requestConfig = RequestConfig.custom().setConnectTimeout(10_000).setSocketTimeout(10_000)
-                    .build();
+            RequestConfig requestConfig = RequestConfig.custom().setConnectionRequestTimeout(Timeout.ofSeconds(10))
+                    .setResponseTimeout(Timeout.ofSeconds(10)).build();
```
Comment (tesshucom) [inline jpsonic-main/src/main/java/com/tesshu/jpsonic/service/VersionService.java:270]: Use SocketTimeout.

File: jpsonic-main/src/main/java/com/tesshu/jpsonic/service/scanner/PodcastServiceImpl.java [hunk 6]
```diff
@@ -336,50 +338,55 @@ private void refreshChannels(final List<PodcastChannel> channels, final boolean
...
+        if (!scannerState.tryScanningLock()) {
+            return;
+        }
+        indexManager.startIndexing();
+
+        channel.setStatus(PodcastStatus.DOWNLOADING);
+        channel.setErrorMessage(null);
+        podcastDao.updateChannel(channel);
+        RequestConfig requestConfig = RequestConfig.custom().setConnectionRequestTimeout(Timeout.ofMinutes(2))
+                .setResponseTimeout(Timeout.ofMinutes(10)).build();
```
Comment (tesshucom) [inline jpsonic-main/src/main/java/com/tesshu/jpsonic/service/scanner/PodcastServiceImpl.java:377]: Use SocketTimeout.

File: jpsonic-main/src/main/java/com/tesshu/jpsonic/service/scanner/PodcastServiceImpl.java [hunk 7]
```diff
@@ -609,15 +628,14 @@ private String getITunesAttribute(Element element, String childName, String attr
     }
 
     private HttpGet createHttpGet(String url) {
-        RequestConfig requestConfig = RequestConfig.custom().setConnectTimeout(2 * 60 * 1000) // 2
-                                                                                              // minutes
-                .setSocketTimeout(10 * 60 * 1000) // 10 minutes
+        RequestConfig requestConfig = RequestConfig.custom().setConnectionRequestTimeout(Timeout.ofMinutes(2))
+                .setResponseTimeout(Timeout.ofMinutes(10))
```
Comment (tesshucom) [inline jpsonic-main/src/main/java/com/tesshu/jpsonic/service/scanner/PodcastServiceImpl.java:632]: Use SocketTimeout.

File: jpsonic-main/src/main/java/com/tesshu/jpsonic/service/scrobbler/LastFMScrobbler.java [hunk 8]
```diff
@@ -64,8 +65,8 @@ public class LastFMScrobbler {
 
     private static final Logger LOG = LoggerFactory.getLogger(LastFMScrobbler.class);
     private static final int MAX_PENDING_REGISTRATION = 2000;
-    private static final RequestConfig REQUEST_CONFIG = RequestConfig.custom().setConnectTimeout(15_000)
-            .setSocketTimeout(15_000).build();
+    private static final RequestConfig REQUEST_CONFIG = RequestConfig.custom()
+            .setConnectionRequestTimeout(Timeout.ofSeconds(15)).setResponseTimeout(Timeout.ofSeconds(15)).build();
```
Comment (tesshucom) [inline jpsonic-main/src/main/java/com/tesshu/jpsonic/service/scrobbler/LastFMScrobbler.java:69]: Use SocketTimeout.

File: jpsonic-main/src/main/java/com/tesshu/jpsonic/service/upnp/ApacheUpnpServiceConfiguration.java [hunk 9]
```diff
@@ -40,17 +30,17 @@
...
+    // }
+    //
+    // @SuppressWarnings("rawtypes")
+    // @Override
+    // public StreamServer createStreamServer(NetworkAddressFactory networkAddressFactory) {
+    // int listenPort = SettingsService.getDefaultUPnPPort();
+    // return new StreamServerImpl(new StreamServerConfigurationImpl(
+    // listenPort < 1 ? networkAddressFactory.getStreamListenPort() : listenPort));
+    // }
 }
```
Comment (tesshucom) [inline jpsonic-main/src/main/java/com/tesshu/jpsonic/service/upnp/ApacheUpnpServiceConfiguration.java:46]: It will be commented out. (Cannot be compiled). Jpsonic cannot handle this, so a patch version of the 3rd party library will be created.

File: repo/org/directwebremoting/dwr-JakartaEE/4.0.0-SNAPSHOT/dwr-JakartaEE-4.0.0-SNAPSHOT-sources.jar
Comment (tesshucom) [inline repo/org/directwebremoting/dwr-JakartaEE/4.0.0-SNAPSHOT/dwr-JakartaEE-4.0.0-SNAPSHOT-sources.jar:1]: Remove repo/org/directwebremoting/dwr/3.0.rc1

Comment (tesshucom) [reply repo/org/directwebremoting/dwr-JakartaEE/4.0.0-SNAPSHOT/dwr-JakartaEE-4.0.0-SNAPSHOT-sources.jar:1]: Also...delete the following.

repo/org/terracotta/ehcache-probe
repo/com/jgoodies/forms

I don't think it's used though 😟
Confirm and delete.

File: jpsonic-main/src/main/java/com/tesshu/jpsonic/service/upnp/ApacheUpnpServiceConfiguration.java (hunk 9 above)
Comment (tesshucom) [reply jpsonic-main/src/main/java/com/tesshu/jpsonic/service/upnp/ApacheUpnpServiceConfiguration.java:46]: Already listed as `Non-Goal`. This issue will be resolved in a subsequent pull request.

File: jpsonic-main/src/main/java/com/tesshu/jpsonic/service/scrobbler/LastFMScrobbler.java (hunk 8 above)
Comment (tesshucom) [reply jpsonic-main/src/main/java/com/tesshu/jpsonic/service/scrobbler/LastFMScrobbler.java:69]: Misunderstanding. OK with ResponseTimeout.

File: airsonic-taglibs/src/main/resources/META-INF/sub.tld (hunk 1 above)
Comment (tesshucom) [reply airsonic-taglibs/src/main/resources/META-INF/sub.tld:6]: Wmm... That's obvious with Java EE 5.

[The Java EE 5 Tutorial - Tag Library Descriptors](https://docs.oracle.com/javaee/5/tutorial/doc/bnamu.html)

Oracle has deprecated JSP technology in JSF, and the content of the 6.0 documentation is ( .. not complete?) and is provided as an external link to JSF. It's a rabbit hole.

https://docs.oracle.com/javaee/6/tutorial/doc/bnawn.html

Since there are no problems so far, it seems that descriptors will not be a big mistake even if they are derived from Java EE 5. Leave it as is for now.

File: repo/org/directwebremoting/dwr-JakartaEE/4.0.0-SNAPSHOT/dwr-JakartaEE-4.0.0-SNAPSHOT-sources.jar
Comment (tesshucom) [reply repo/org/directwebremoting/dwr-JakartaEE/4.0.0-SNAPSHOT/dwr-JakartaEE-4.0.0-SNAPSHOT-sources.jar:1]: Done

Comment (tesshucom) [reply repo/org/directwebremoting/dwr-JakartaEE/4.0.0-SNAPSHOT/dwr-JakartaEE-4.0.0-SNAPSHOT-sources.jar:1]: It looks like natpmp is being used. However, the UPnP implementation will change in a later pull request. At the moment, that is.

Comment (tesshucom) [reply repo/org/directwebremoting/dwr-JakartaEE/4.0.0-SNAPSHOT/dwr-JakartaEE-4.0.0-SNAPSHOT-sources.jar:1]: Done. If there's any problem, I'll deal with it then 🤪
